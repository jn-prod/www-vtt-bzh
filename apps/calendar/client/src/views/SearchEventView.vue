<template>
  <section class="container-fluid">
    <header id="header-wrapper" class="row mb-5">
      <section class="col-md-4 bg-white p-2 m-2 mx-md-5 px-md-5 rounded shadow">
        <!-- value proposition        -->
        <h1 class="h3 text-left m2-5">Rechercher une rando VTT à coté de chez toi n'aura jamais été aussi simple.</h1>

        <!-- search form component -->
        <search-form @submit-search-form="setSearchQuery" @cancel-search-form="cancelSearchQuery"></search-form>
      </section>
    </header>

    <!-- calendar component -->
    <div class="mt-5 mx-md-5">
      <h2 id="calendar" class="mb-5 text-center">
        Calendrier des randonnées à venir
        <span id="nombre_rando" class="badge bg-success">{{ count }}</span>
      </h2>
      <p v-if="isResults" class="alert alert-danger">
        Aucun résultat pour cette recherche, choisissez une autre date de début et de fin.
      </p>

      <!-- Load event -->
      <event-card v-for="event in events" :key="event.id" :event="event" class="my-3"> </event-card>

      <!-- load-more button -->
      <div class="text-center mt-3">
        <button
          v-show="isLoadMoreActive"
          class="btn btn-secondary shadow mt-3 px-5 rounded-pill"
          @click.prevent="setPaginator()"
        >
          Voir +
        </button>
      </div>
    </div>
  </section>
</template>

<script lang="js">
import { computed, ref } from 'vue';

import eventService from '@/services/events.service';

// import components
import EventCard from '@/components/EventCard.vue';
import SearchForm from '@/components/SearchForm.vue';

const getDate = (date) => date.toISOString().split('T')[0];

export default {
  name: 'SearchEventView',
  components: {
    EventCard,
    SearchForm,
  },
  setup() {
    const data = ref([]);
    const paginator = ref(20);
    const searchFormQuery = ref(null);
    const baseQuery = { from: `${getDate(new Date())}` };

    (async () => {
      const { data: events } = await eventService.fetchSearch(baseQuery);
      data.value = events;
    })();

    const setPaginator = (inc = 20) => {
      paginator.value += inc;
    };

    const setSearchQuery = async ({ dpt, endDate, startDate }) => {
      if (!dpt && !endDate && !startDate) return;

      searchFormQuery.value = {
        where: dpt === 'all' ? undefined : dpt,
        from: getDate(startDate),
        to: getDate(endDate),
      };

      const { data: events } = await eventService.fetchSearch(searchFormQuery.value);
      data.value = events;
    };

    const cancelSearchQuery = async ({ cancel }) => {
      if (!cancel) return;

      searchFormQuery.value = baseQuery;
      const { data: events } = await eventService.fetchSearch(baseQuery);
      data.value = events;
    };

    const isLoadMoreActive = computed(() => (data.value || []).length > (paginator.value || 0));

    const events = computed(() => (data.value || []).slice(0, paginator.value));

    const isResults = computed(() => data.value?.length < 1);

    const count = computed(() => data?.value?.length || 0);

    return {
      setSearchQuery,
      cancelSearchQuery,
      setPaginator,
      isLoadMoreActive,
      events,
      count,
      isResults,
    };
  },
};
</script>
