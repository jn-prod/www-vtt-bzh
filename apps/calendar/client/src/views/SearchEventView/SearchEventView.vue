<template>
  <section class="container-fluid">
    <!-- calendar component -->
    <div class="mt-5 mx-md-5">
      <h1 id="calendar" class="mb-5 text-center">
        Calendrier
      </h1>

      <!-- search form component -->
      <search-form @submit-search-form="setSearchQuery" @cancel-search-form="cancelSearchQuery"></search-form>

      <p v-if="!isResults && !isLoading" class="alert alert-danger" aria-live="polite" :aria-busy="isLoading">
        Aucun résultat pour cette recherche, choisissez une autre date de début et de fin.
      </p>
      <p v-if="isResults && !isLoading" class="alert alert-info" aria-live="polite" :aria-busy="isLoading">
        {{ count }} randonnées à découvrir.
      </p>
      <p v-if="isLoading" class="alert alert-info" aria-live="polite" :aria-busy="isLoading">Chargement en cours.</p>

      <!-- Load event -->
      <event-card v-for="event in events" :key="event.id" :event="event" class="my-3"> </event-card>

      <!-- load-more button -->
      <div class="text-center mt-3">
        <button
          v-show="isLoadMoreActive"
          class="btn btn-secondary shadow mt-3 px-5 rounded-pill"
          @click.prevent="setPaginator()"
        >
          Voir +
        </button>
      </div>
    </div>
  </section>
</template>

<script lang="js">
import { computed, ref } from 'vue';

import eventService from '@/services/events.service';

// import components
import EventCard from '@/components/EventCard.vue';
import SearchForm from '@/components/SearchForm.vue';

const getDate = (date) => date.toISOString().split('T')[0];

export default {
  name: 'SearchEventView',
  components: {
    EventCard,
    SearchForm,
  },
  setup() {
    const data = ref([]);
    const paginator = ref(20);
    const searchFormQuery = ref(null);
    const isLoading = ref(true);
    const baseQuery = { from: `${getDate(new Date())}` };

    const search = async (query) => {
      isLoading.value = true;
      try {
        data.value = await eventService.fetchSearch(query);
      } catch (err) {
        console.error(err);
      }
      isLoading.value = false;
    };

    (async () => {
      await search(baseQuery);
    })();

    const setPaginator = (inc = 20) => {
      paginator.value += inc;
    };

    const setSearchQuery = async ({ dpt, endDate, startDate }) => {
      if (!dpt && !endDate && !startDate) return;

      searchFormQuery.value = {
        where: dpt === 'all' ? undefined : dpt,
        from: getDate(startDate),
        to: getDate(endDate),
      };

      await search(searchFormQuery.value);
    };

    const cancelSearchQuery = async ({ cancel }) => {
      if (!cancel) return;

      searchFormQuery.value = baseQuery;
      await search(baseQuery);
    };

    const count = computed(() => data?.value?.length || 0);

    const isLoadMoreActive = computed(() => count.value > (paginator.value || 0));

    const events = computed(() => (data.value || []).slice(0, paginator.value));

    const isResults = computed(() => count.value > 0);

    return {
      setSearchQuery,
      cancelSearchQuery,
      setPaginator,
      isLoadMoreActive,
      events,
      count,
      isResults,
      isLoading,
    };
  },
};
</script>
