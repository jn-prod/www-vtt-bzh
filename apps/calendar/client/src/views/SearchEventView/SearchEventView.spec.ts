import { describe, expect, test } from 'vitest';
import { mount } from '@vue/test-utils';
// import { axe, toHaveNoViolations } from 'jest-axe';
import SearchEventView from './SearchEventView.vue';

const createWrapper = () => mount(SearchEventView);

describe('SearchEventView', () => {
  test('render', async () => {
    const vm = createWrapper();
    // expect(await axe(vm.html())).toHaveNoViolations();
    expect(vm.html()).toMatchSnapshot();
  });
});
