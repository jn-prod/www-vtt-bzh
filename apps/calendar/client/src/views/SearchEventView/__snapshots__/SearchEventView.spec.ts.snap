// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchEventView > render 1`] = `
<section class="container-fluid">
  <!-- calendar component -->
  <div class="mt-5 mx-md-5">
    <h1 class="mb-5 text-center" id="calendar"> Calendrier des randonnées à venir <span class="badge bg-success" id="nombre_rando">0</span>
    </h1><!-- search form component -->
    <article class="my-2">
      <header class="d-flex">
        <button aria-controls="search" class="btn btn-light ms-auto">
          <!--v-if-->
          <span class="my-auto me-2">
            <i aria-hidden="true" class="fas fa-chevron-down"></i>
          </span> Filtrer
        </button>
      </header>
      <form aria-expanded="false" hidden id="search">
        <div class="my-2">
          <span>
            <label for="start-date">Debut</label>
            <input class="form-control" id="start-date" name="start-date" type="date">
          </span>
        </div>
        <div class="my-2">
          <span>
            <label for="end-date">Fin</label>
            <input class="form-control" id="end-date" name="end-date" type="date">
          </span>
        </div>
        <div class="my-2">
          <label for="departement"> Département </label>
          <select class="form-control" id="departement" name="departement">
            <option value="22">Côtes d'Armor</option>
            <option value="29">Finistère</option>
            <option value="35">Ille et Vilaine</option>
            <option value="44">Loire Atlantique</option>
            <option value="56">Morbihan</option>
            <option value="all">Tous les départements</option>
          </select>
        </div>
        <div class="my-2">
          <div class="mx-auto">
            <button class="btn btn-secondary shadow m-2 rounded-pill" id="search_button" type="submit">
              <i class="fas fa-search"></i> Rechercher
            </button>
            <button class="btn btn-outline-dark rounded-pill shadow m-2" href="#calendar">
              <i class="far fa-trash-alt"></i> Réinitialiser
            </button>
          </div>
        </div>
      </form>
    </article>
    <!--v-if-->
    <p aria-busy="true" aria-live="polite" class="alert alert-info">Chargement en cours.</p><!-- Load event -->
    <!-- load-more button -->
    <div class="text-center mt-3">
      <button class="btn btn-secondary shadow mt-3 px-5 rounded-pill" style="display: none;"> Voir + </button>
    </div>
  </div>
</section>
`;
